File: src/MyServiceBus.Abstractions/PipeConfigurator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;

namespace MyServiceBus;

public class PipeConfigurator<TContext>
    where TContext : class, PipeContext
{
    readonly List<FilterRegistration> filters = new();

    public void UseFilter(IFilter<TContext> filter)
    {
        AddFilter(
            _ => filter,
            "filter",
            filter.GetType(),
            FilterLifetime.Instance);
    }

    public void UseFilter<TFilter>()
        where TFilter : class, IFilter<TContext>
    {
        AddFilter(
            provider => provider != null
                ? (IFilter<TContext>)ActivatorUtilities.GetServiceOrCreateInstance(provider, typeof(TFilter))
                : Activator.CreateInstance<TFilter>()!,
            "filter",
            typeof(TFilter),
            FilterLifetime.Pipe);
    }

    public void UseScopedFilter<TFilter>()
        where TFilter : class, IFilter<TContext>
    {
        AddFilter(
            provider => provider != null
                ? new ScopedFilter<TFilter>(provider)
                : throw new InvalidOperationException("A service provider is required to use a scoped filter"),
            "filter",
            typeof(TFilter),
            FilterLifetime.Scoped);
    }

    public void UseExecute(Func<TContext, Task> callback)
    {
        AddFilter(
            _ => new DelegateFilter(callback),
            "execute",
            null,
            FilterLifetime.Instance);
    }

    public void UseRetry(int retryCount, TimeSpan? delay = null)
    {
        if (retryCount < 0)
            throw new ArgumentOutOfRangeException(nameof(retryCount));

        var configuration = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["retryCount"] = retryCount.ToString(CultureInfo.InvariantCulture)
        };
        if (delay.HasValue)
            configuration["delayMilliseconds"] = delay.Value.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);

        AddFilter(
            _ => new RetryFilter<TContext>(retryCount, delay),
            "retry",
            typeof(RetryFilter<TContext>),
            FilterLifetime.Pipe,
            configuration);
    }

    public void UseMessageRetry(Action<RetryConfigurator> configure)
    {
        var rc = new RetryConfigurator();
        configure(rc);
        UseRetry(rc.RetryCount, rc.Delay);
    }

    public IPipe<TContext> Build(IServiceProvider? provider = null)
    {
        IPipe<TContext> next = Pipe.Empty<TContext>();
        for (var i = filters.Count - 1; i >= 0; i--)
        {
            var filter = filters[i].Factory(provider);
            next = new FilterPipe(filter, next);
        }

        return next;
    }

    public PipelineDescriptor GetDescriptor()
    {
        return new PipelineDescriptor(
            filters.Select(x => x.Descriptor).ToArray());
    }

    void AddFilter(
        Func<IServiceProvider?, IFilter<TContext>> factory,
        string kind,
        Type? implementation,
        FilterLifetime lifetime,
        IReadOnlyDictionary<string, string>? configuration = null)
    {
        filters.Add(new FilterRegistration(
            factory,
            new FilterDescriptor(
                filters.Count,
                kind,
                implementation?.FullName,
                lifetime,
                configuration)));
    }

    sealed record FilterRegistration(
        Func<IServiceProvider?, IFilter<TContext>> Factory,
        FilterDescriptor Descriptor);

    class DelegateFilter : IFilter<TContext>
    {
        readonly Func<TContext, Task> callback;

        public DelegateFilter(Func<TContext, Task> callback)
        {
            this.callback = callback;
        }

        public async Task Send(TContext context, IPipe<TContext> next)
        {
            await callback(context);
            await next.Send(context);
        }
    }

    class FilterPipe : IPipe<TContext>
    {
        readonly IFilter<TContext> filter;
        readonly IPipe<TContext> next;

        public FilterPipe(IFilter<TContext> filter, IPipe<TContext> next)
        {
            this.filter = filter;
            this.next = next;
        }

        public Task Send(TContext context)
        {
            return filter.Send(context, next);
        }
    }

    sealed class ScopedFilter<TFilter> : IFilter<TContext>
        where TFilter : class, IFilter<TContext>
    {
        readonly IServiceProvider provider;

        public ScopedFilter(IServiceProvider provider)
        {
            this.provider = provider;
        }

        public async Task Send(TContext context, IPipe<TContext> next)
        {
            await using var scope = provider.CreateAsyncScope();
            var filter = scope.ServiceProvider.GetRequiredService<TFilter>();
            await filter.Send(context, next);
        }
    }

}

File: src/Java/myservicebus-abstractions/src/main/java/com/myservicebus/PipeConfigurator.java
package com.myservicebus;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.Map;

import com.myservicebus.di.ServiceProvider;
import com.myservicebus.di.ServiceScope;

/**
 * Configures and builds pipes by chaining filters.
 */
public class PipeConfigurator<TContext extends PipeContext> {
    private final List<FilterRegistration<TContext>> filters = new ArrayList<>();

    public void useFilter(Filter<TContext> filter) {
        addFilter(sp -> filter, "filter", filter.getClass(), FilterLifetime.INSTANCE, Map.of());
    }

    public void useFilter(Class<? extends Filter<TContext>> filterClass) {
        addFilter(sp -> {
            try {
                if (sp != null) {
                    Filter<TContext> resolved = sp.getService(filterClass);
                    if (resolved != null) {
                        return resolved;
                    }
                }
                return filterClass.getDeclaredConstructor().newInstance();
            } catch (Exception ex) {
                throw new RuntimeException("Failed to create filter " + filterClass.getName(), ex);
            }
        }, "filter", filterClass, FilterLifetime.PIPE, Map.of());
    }

    public void useScopedFilter(Class<? extends Filter<TContext>> filterClass) {
        addFilter(provider -> {
            if (provider == null) {
                throw new IllegalStateException("A service provider is required to use a scoped filter");
            }
            return (context, next) -> {
                ServiceScope scope = provider.createScope();
                try {
                    Filter<TContext> filter = providerFor(scope).getRequiredService(filterClass);
                    CompletableFuture<Void> result = filter.send(context, next);
                    scope.detach();
                    return result.whenComplete((ignored, failure) -> scope.close());
                } catch (Throwable failure) {
                    scope.close();
                    return CompletableFuture.failedFuture(failure);
                }
            };
        }, "filter", filterClass, FilterLifetime.SCOPED, Map.of());
    }

    private static ServiceProvider providerFor(ServiceScope scope) {
        return scope.getServiceProvider();
    }

    public void useExecute(Function<TContext, CompletableFuture<Void>> callback) {
        addFilter(sp -> new DelegateFilter(callback), "execute", null, FilterLifetime.INSTANCE, Map.of());
    }

    public void useRetry(int retryCount) {
        useRetry(retryCount, null);
    }

    public void useRetry(int retryCount, Duration delay) {
        if (retryCount < 0) {
            throw new IllegalArgumentException("retryCount");
        }
        Map<String, String> configuration = delay == null
                ? Map.of("retryCount", Integer.toString(retryCount))
                : Map.of(
                        "retryCount", Integer.toString(retryCount),
                        "delayMilliseconds", Long.toString(delay.toMillis()));
        addFilter(
                sp -> new RetryFilter<>(retryCount, delay),
                "retry",
                RetryFilter.class,
                FilterLifetime.PIPE,
                configuration);
    }

    public void useMessageRetry(java.util.function.Consumer<RetryConfigurator> configure) {
        RetryConfigurator rc = new RetryConfigurator();
        configure.accept(rc);
        useRetry(rc.getRetryCount(), rc.getDelay());
    }

    public Pipe<TContext> build() {
        return build(null);
    }

    public Pipe<TContext> build(ServiceProvider provider) {
        Pipe<TContext> next = Pipes.empty();
        for (int i = filters.size() - 1; i >= 0; i--) {
            Filter<TContext> filter = filters.get(i).factory().apply(provider);
            Pipe<TContext> current = next;
            next = ctx -> filter.send(ctx, current);
        }
        return next;
    }

    public PipelineDescriptor getDescriptor() {
        return new PipelineDescriptor(
                filters.stream().map(FilterRegistration::descriptor).toList());
    }

    private void addFilter(
            Function<ServiceProvider, Filter<TContext>> factory,
            String kind,
            Class<?> implementation,
            FilterLifetime lifetime,
            Map<String, String> configuration) {
        filters.add(new FilterRegistration<>(
                factory,
                new FilterDescriptor(
                        filters.size(),
                        kind,
                        implementation != null ? implementation.getName() : null,
                        lifetime,
                        configuration)));
    }

    private record FilterRegistration<TContext extends PipeContext>(
            Function<ServiceProvider, Filter<TContext>> factory,
            FilterDescriptor descriptor) {
    }

    class DelegateFilter implements Filter<TContext> {
        private final Function<TContext, CompletableFuture<Void>> callback;

        DelegateFilter(Function<TContext, CompletableFuture<Void>> callback) {
            this.callback = callback;
        }

        @Override
        public CompletableFuture<Void> send(TContext context, Pipe<TContext> next) {
            return callback.apply(context).thenCompose(v -> next.send(context));
        }
    }
}
